### ISSUE TITLE
switching layouts changes buffer numbers when neo tree buffer is open

### ISSUE THREAD

**Issue by paradigmshift** [syl20bnr/spacemacs#4710]
The Neo Tree buffer by default is buffer 0 (m-0). Now if I switch to another layout and then go back to the layout with the Neo Tree buffer, all the buffer numbers are incremented by 1 (Neo Tree becomes 1, other buffers n + 1). Another quirk is that if I click on an un-opened folder in Neo Tree after this happens, another Neo Tree buffer pops open with a buffer number of 0, so I end up with two Neo Tree buffers (0 and 1).

![uuqjahe - imgur](https://cloud.githubusercontent.com/assets/846449/12462641/b4170c76-bfbf-11e5-813c-c4b3ac9c9cb3.gif)

**Comment by bmag** [comment]
#4820 should fix the bug that causes neotree's window number to change.

The "another quirk" you mentioned is an entirely different bug: when switching back and forth between layouts, neotree loses track of its window. It opens a new neotree window because it "forgets" that it already has a window. That should be fixed either in neotree upstream (if possible), or by some glue to integrate neotree and persp-mode together.

I didn't check, but neotree might forget its window also when switching back and forth between workspaces. If that's so, then we also need some glue to integrate neotree and eyebrowse.

**Comment by bmag** [comment]
> when switching back and forth between layouts, neotree loses track of its window

I've debugged this some more, here are my findings:

TL;DR: we need to write a wrapper function with a guard around `neo-global--attach` and add it to the hooks `persp-activated-hook` and `eyebrowse-post-window-switch-hook`.

When switching layouts (`SPC l <number>`) or workspaces (`SPC l w <number>`), neotree loses track of its window. I didn't get to the bottom of how exactly this happens, but its likely because neotree saves a window object as a way to keep track of its window (see `neo-global--window`). Switching back-and-forth between layouts/workspaces restores the window configuration, but doesn't actually use the same window objects, so technically neotree's window has changed although it looks exactly the same as the old one (that's saved in `neo-global--window`).

Neotree has a function called `neo-global--attach` that updates the `neo-global--window` object, and there's actually some code in neotree to call `neo-global--attach` after popwin creates its popup window (popwin creates the same problem for neotree as switching layouts/workspaces).

So in short, we need to hook `neo-global--attach` into `persp-activated-hook` (runs after switching a layout) and `eyebrowse-post-window-switch-hook` (runs after switching a workspace).

However, I'm not sure what problems it may cause to people who use spacemacs with multiple frames. In their case, they can have neotree open in one frame, and switch the workspace of another frame (workspaces are frame-local). Calling `neo-global--attach` in this situation will make neotree forget about its window in the first frame. I'll try to come up with a sane guard to put around `neo-global--attach` that we can hook into the appropriate places.

**Comment by syl20bnr** [comment]
@bmag Thank you for the detailed report. We need to report this upstream, seems that have several issues for Neotree:
1) integration with perps and eyebrowse
2) multi-frame support.

**Comment by syl20bnr** [comment]
For reference:

> Note: requires nschum/window-numbering.el#11 to be merged upstream into window-numbering package.

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.
